core/src/main/java/io/grpc/ChannelImpl.java

```diff
@@ -356,10 +370,23 @@ public boolean isReady() {
     /**
      * Set missing properties on the headers. The given headers will be mutated.
      * @param headers the headers to complete
+     * @param timeoutMicros timeout in microseconds. -1 for infinite. Otherwise, must &gt; 0
      */
-    private void completeHeaders(Metadata.Headers headers) {
+    private void completeHeaders(Metadata.Headers headers, long timeoutMicros) {
```

Comment:
I need to return early in ``start()`` if the deadline is already exceeded, so the timeout must be calculated in ``start()``, not in a down call. I find it nicer to just move all the ``completeHeaders()`` stuff to ``start()``.